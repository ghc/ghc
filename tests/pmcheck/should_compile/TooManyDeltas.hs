-- | The function here exploit matches of arity 2 that split the uncovered set
-- in two. Too many for -fmax-pmcheck-models=0!
-- As a result, these functions elicit the symptoms describe in the warnings
-- message, e.g.
--   - False positives on exhaustivity
--   - Turns redundant into inaccessible clauses
--   - Fails to report redundant matches
module TooManyDeltas where

data T = A | B

-- | Reports that a clause for _ _ is missing.
f :: T -> T -> ()
f A A = ()

-- | Reports that the third clause is inaccessible, when really it is
-- redundant.
g :: T -> T -> ()
g _ A = ()
g A A = () -- inaccessible, correctly flagged
g A A = () -- redundant, not inaccessible!
g _ _ = () -- (this one is not about exhaustivity)

-- | Fails to report that the second clause is redundant.
h :: T -> T -> ()
h A A = () -- covered, emits no warning
h A A = () -- redundant, not covered!
h _ _ = () -- (this one is not about exhaustivity)
