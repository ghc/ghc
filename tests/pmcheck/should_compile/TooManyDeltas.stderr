
TooManyDeltas.hs:14:1: warning: [GHC-62161] [-Wincomplete-patterns (in -Wextra)]
    Pattern match(es) are non-exhaustive
    In an equation for ‘f’: Patterns of type ‘T’, ‘T’ not matched: _ _

TooManyDeltas.hs:14:1: warning: [GHC-61505]
    Pattern match checker ran into -fmax-pmcheck-models=0 limit, so
      • Redundant clauses might not be reported at all
      • Redundant clauses might be reported as inaccessible
      • Patterns reported as unmatched might actually be matched
    Suggested fix:
      Increase the limit or resolve the warnings to suppress this message.

TooManyDeltas.hs:19:1: warning: [GHC-61505]
    Pattern match checker ran into -fmax-pmcheck-models=0 limit, so
      • Redundant clauses might not be reported at all
      • Redundant clauses might be reported as inaccessible
      • Patterns reported as unmatched might actually be matched
    Suggested fix:
      Increase the limit or resolve the warnings to suppress this message.

TooManyDeltas.hs:20:1: warning: [GHC-94210] [-Woverlapping-patterns (in -Wdefault)]
    Pattern match has inaccessible right hand side
    In an equation for ‘g’: g A A = ...

TooManyDeltas.hs:21:1: warning: [GHC-94210] [-Woverlapping-patterns (in -Wdefault)]
    Pattern match has inaccessible right hand side
    In an equation for ‘g’: g A A = ...
